### DISCUSSION TITLE
How do the limit switches work? (Ramps 1.6, Due, GrblHAL, UGS)

### DISCUSSION THREAD

**Discussion by Racaya** [terjeio/grblHAL#264] [Q&A]
Hello all,

Currently I am trying to get a Ramps 1.6 with an Arduino Due to move around some stepper motors. I am using GrblHAL test branch and UGS for communication. However, I got some problems with the limit switches. I will post the same question in the UGS git since I have no clue where the problem may be lying. 

Using [the typical switches](https://www.banggood.com/Geekcreit-RAMPS-1_4-Endstop-Switch-For-RepRap-Mendel-3D-Printer-With-70cm-Cable-p-958124.html?cur_warehouse=UK) I seem to be unable to get the controller do something useful.

Whenever I press the button, nothing happens. Only when I release, the controller goes into Alarm state. Same if I invert the limit switches. Same behaviour when I short the signal pin to either - or +.

I tried using the UGS setup wizard. There in the limit switches section, the green fields should turn red when the limit switches are pressed. Unfortunately, that does not happen. The controller goes into Alarm state upon release, but the the green field never changes. Here again it doesn't matter whether I invert the switches or not.
The other limit switches can be shorted between signal and - or + or left open, no difference either. 

![Before..](https://user-images.githubusercontent.com/14227479/108748178-96998e00-753e-11eb-8075-1a529cadd57d.png)

![.. after pressing and releasing the limit switch](https://user-images.githubusercontent.com/14227479/108748230-a4e7aa00-753e-11eb-8357-0f02c50c37e9.png)

I tried looking into the firmware, but looks like my C skills are insufficient to understand the inner workings of grblHAL ;(

So, how do the limit switches work? Am I doing something wrong with the cabling? Do they all need to be in a certain state before setup? 

Thank you in advance!

**Comment by terjeio** [comment]
What type of switches (normally open or normally closed) and how many?

Ramps 1.6 has two limit inputs per axis, min & max. If using normally closed switches (which is the grblHAL default) connected to the min inputs only the max inputs has to be shorted to ground for correct operation.

If normally open then this is not neccesary - but the $5 has to be set to 7 to invert all inputs for a 3-axis machine. It seems have made a mistake with this as legacy grbl only allows 0 or 1 - grblHAL with [compatibility level](https://github.com/terjeio/grblHAL/wiki/Compatibility-level) set > 1 should behave the same but does not. From the images above it looks like UGS sets $5 to either 0 or 1 - when 1 only X is inverted.

**Reply by Racaya** [reply]
> 
> What type of switches (normally open or normally closed) and how many?
> 
> Ramps 1.6 has two limit inputs per axis, min & max. If using normally closed switches (which is the grblHAL default) connected to the min inputs only the max inputs has to be shorted to ground for correct operation.
>

Currently only one switch, which is NC between signal and high/+ and NO between signal and low/-. 
From your answer I guess that's the way it should be (from grbl's perspective, NC means that there's normally a connection to high, right?). But shouldn't I then short the others to high? Because if shorted to ground that means that the switch is actually "pressed" and therefore the limit triggers?

> 
> If normally open then this is not neccesary - but the $5 has to be set to 7 to invert all inputs for a 3-axis machine. It seems have made a mistake with this as legacy grbl only allows 0 or 1 - grblHAL with [compatibility level](https://github.com/terjeio/grblHAL/wiki/Compatibility-level) set > 1 should behave the same but does not. From the images above it looks like UGS sets $5 to either 0 or 1 - when 1 only X is inverted.

Setting $5=7 actually changes the behaviour so that the Alarm state gets triggered the moment I press the button (i.e. shorting signal to gnd/low) and not, like before, the moment I release the button.

**Reply by Racaya** [reply]
After playing around a bit with different variations of $5 and shorting the other axes I got the feeling that it does no difference at all how I short them. The alarm only gets triggered when there is a change in state, not depending on the current state.

**Reply by Racaya** [reply]
Next update:
After setting $10 to 121 like recommended bei @breiler I can finally see the alarm state correctly. 

What is still confusing me: 
With $5=0, the limit signal pins need to be connected to GND for no alarm.
With $5=7, the limit signal pins need to be connected to HIGH for no alarm.

Shouldn't it be the other way round? Normally closed meaning that normally there is HIGH on signal? Like, if the cable breaks there is no connection, meaning signal floats which is more like being connected to GND and the alarm triggers? 
I can handle both and design my electronics accordingly. Just got the feeling that either I miss something here or there is some error in grblHAL (which I don't expect, but better be safe than sorry ;) )

**Reply by terjeio** [reply]
> Shouldn't it be the other way round?

No, in the grblHAL default configuration a switch is connected between the input (which is pulled up) and ground. With a normally closed switch this means the input is normally low and goes high when the switch opens or the wire breaks.

Legacy grbl defaults to pulled up inputs and normally open switches which is bad as a broken wire will not be detected. Noise immunity is not as good in this configuration either as the cable from the switch to the pin is acting as an antenna with only the pullup resistor as the load. If the pullup resistor is internal to the MCU then this is typically in the 50Kohm range. With a normally closed switch the load will be significantly higher (lower impedance) as the the wire is grounded via the switch.

**Comment by breiler** [comment]
What is your settings (from $$) and compatibility mode?

There is now a bitmask in GrblHAL that controls what information is included in the status report. It could be that the pin state is not being reported? See this [page](https://github.com/terjeio/grblHAL/wiki/Additional-or-extended-settings---Copy) about the $10-setting.

As @terjeio mentioned, if you are using compatibility mode higher than zero the settings for inverting the limit switches has changed in GrblHAL and is not supported in UGS. The pin state should however be recognized if it has been enabled in the $10-settings.

**Reply by Racaya** [reply]
This the output of $$ ($5 is now 7, but I also tried with 0 and 1):

> $$
$0 = 10.0    (Step pulse time, microseconds)
$1 = 255    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 3    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 7    (Invert limit pins, boolean)
$6 = 0    (Invert probe pin, boolean)
$7 = 0   
$10 = 1    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$14 = 7   
$15 = 0   
$16 = 0   
$17 = 0   
$18 = 0   
$19 = 0   
$20 = 0    (Soft limits enable, boolean)
$21 = 1    (Hard limits enable, boolean)
$22 = 0    (Homing cycle enable, boolean)
$23 = 0    (Homing direction invert, mask)
$24 = 25.0    (Homing locate feed rate, mm/min)
$25 = 500.0    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 1.000    (Homing switch pull-off distance, millimeters)
$28 = 0.100   
$29 = 0.0   
$30 = 256.000    (Maximum spindle speed, RPM)
$31 = 0.000    (Minimum spindle speed, RPM)
$32 = 0    (Laser-mode enable, boolean)
$33 = 5000.0   
$34 = 0.0   
$35 = 0.0   
$36 = 100.0   
$37 = 0   
$39 = 0   
$40 = 0   
$43 = 1   
$44 = 4   
$45 = 3   
$46 = 0   
$62 = 0   
$63 = 2   
$64 = 0   
$65 = 0   
$80 = 1.000   
$81 = 0.010   
$82 = 0.000   
$84 = 0.000   
$85 = 10.000   
$90 = 0.000   
$91 = 0.000   
$92 = 0.000   
$95 = 0.000   
$100 = 23.600    (X-axis travel resolution, step/mm)
$101 = 13.400    (Y-axis travel resolution, step/mm)
$102 = 8.889    (Z-axis travel resolution, step/mm)
$110 = 7000.000    (X-axis maximum rate, mm/min)
$111 = 5000.000    (Y-axis maximum rate, mm/min)
$112 = 5000.000    (Z-axis maximum rate, mm/min)
$120 = 150.000    (X-axis acceleration, mm/sec^2)
$121 = 150.000    (Y-axis acceleration, mm/sec^2)
$122 = 50.000    (Z-axis acceleration, mm/sec^2)
$130 = 400.000    (X-axis maximum travel, millimeters)
$131 = 2300.000    (Y-axis maximum travel, millimeters)
$132 = 360.000    (Z-axis maximum travel, millimeters)
$341 = 0   
$342 = 30.0   
$343 = 25.0   
$344 = 200.0   
$345 = 100.0   
$347 = 5.0   
$348 = 2.500   
$349 = 25.000   
ok

**Reply by breiler** [reply]
Your setting $10 is set to 1 which looks like it would only report the machine position (@terjeio correct me if I'm wrong):
> 0 = machine_position, 1 = buffer_state, 2 = line_numbers, 3 = feed_speed, 4 = pin_state, 5 = work_coord_offset, 6 = overrides, 7 = probe_coordinates, 8 = sync_on_wco_change, 9 = parser_state, 10 = alarm_substate, 11 - run substate

Without knowing exactly what each of these means, I would probably want to enable 0, 3, 4, 5, 6 to be able to use most functions in UGS (1 + 8 + 16 + 32 + 64 = 121).

**Reply by Racaya** [reply]
I just tried setting $10=121 - seems to work like a charm! Well, at least the interface showing the alarm state. Thank you!

**Reply by terjeio** [reply]
> Your setting $10 is set to 1 which looks like it would only report the machine position (@terjeio correct me if I'm wrong):

You are right - if comaptibility level <= 1.

If > 1 then only bit 0 and 1 is settable. All are reported when > 1 which is another minor bug to be fixed.
The other bits are set via compile time options in _config.h_ and are not overridable unless flashed with comaptibility level <= 1, reconfigured and then reflashed with compatibility level > 1 _without_ resetting settings to default with `$RST=*`. 

https://github.com/terjeio/grblHAL/blob/f85da79de95db12eff347f146f862bdfbb4f683d/drivers/SAM3X8E/grblHAL_Due/src/grbl/config.h#L446-L469

Due to the $10 report bug we can the be fairly sure that the current firmware is compiled with compatibility level <= 1 and thus no limit switch status is reported.

For info: later versions has help available for settings, e.g. `$$=10` issued when compatibility level > 1 outputs this:

```
$10: Status report options as bitfield:
    0 - Position in machine coordinate (1)
    1 - Buffer state (2)
```

There is also a command, `$LIM`, that has recently been added that outputs the switch states independently when there are more than one input per axis. Example with noninverting inputs where only Y min is shorted to ground:

`[LIMITS:XZ,XYZ,,]`

This command can be used to check wiring regardless of the $10 flag setting for pin state report.

> Currently only one switch, which is NC between signal and high/+ and NO between signal and low/-.

This means you have to invert the inputs with `$5=7`.
